polkadot/node/network/collator-protocol/src/validator_side_experimental/collation_manager/mod.rs

```diff
@@ -0,0 +1,1361 @@
+		if let Some(removed) = self.peer_advertisements.remove(peer_id) {
+			for advertisement in removed.advertisements {
+				self.advertisement_timestamps.remove(&advertisement);
+			}
+		}
+	}
+}
+
+#[derive(Default)]
+struct PeerAdvertisements {
+	advertisements: HashSet<Advertisement>,
+	// We increment this even for advertisements that we don't end up accepting, so that we take
+	// these into account when rate limiting.
+	total: usize,
+}
+
+struct PerSessionInfo {
+	our_group: Option<GroupIndex>,
+	n_cores: usize,
+	// The group rotation info changes once per session, apart from the `now` field. The caller
+	// must ensure to override it with the right value.
+	group_rotation_info: GroupRotationInfo,
+	v2_receipts: bool,
+}
+
+// Requests backing subsystem to sanity check the advertisement.
+async fn backing_allows_seconding<Sender>(
+	sender: &mut Sender,
+	advertisement: &Advertisement,
+) -> bool
+where
+	Sender: CollatorProtocolSenderTrait,
+{
+	let Some(prospective_candidate) = advertisement.prospective_candidate else {
+		// Nothing to check for v1 protocol.
+		return true
+	};
+
+	let request = CanSecondRequest {
+		candidate_para_id: advertisement.para_id,
+		candidate_relay_parent: advertisement.relay_parent,
+		candidate_hash: prospective_candidate.candidate_hash,
+		parent_head_data_hash: prospective_candidate.parent_head_data_hash,
+	};
+	let (tx, rx) = oneshot::channel();
+	sender.send_message(CandidateBackingMessage::CanSecond(request, tx)).await;
+
+	rx.await.unwrap_or_else(|err| {
+		gum::warn!(
+			target: LOG_TARGET,
+			?err,
+			relay_parent = ?advertisement.relay_parent,
+			para_id = ?advertisement.para_id,
+			candidate_hash = ?prospective_candidate.candidate_hash,
+			"CanSecond-request responder was dropped",
+		);
+
+		false
+	})
+}
+
+async fn fetch_pvd<Sender: CollatorProtocolSenderTrait>(
+	sender: &mut Sender,
+	receipt: &CandidateReceipt,
+	maybe_parent_head_data_hash: Option<Hash>,
+	maybe_parent_head_data: Option<HeadData>,
+) -> std::result::Result<PersistedValidationData, SecondingError> {
+	let para_id = receipt.descriptor.para_id();
+
+	let pvd = match maybe_parent_head_data_hash {
+		Some(parent_head_data_hash) => {
+			let maybe_pvd = request_prospective_validation_data(
+				sender,
+				receipt.descriptor.relay_parent(),
+				parent_head_data_hash,
+				para_id,
+				maybe_parent_head_data.clone(),
+			)
+			.await?;
+
+			let pvd = match (maybe_pvd, &maybe_parent_head_data) {
+				(None, None) => return Err(SecondingError::BlockedOnParent(parent_head_data_hash)),
+				(Some(pvd), None) => {
+					if parent_head_data_hash != pvd.parent_head.hash() {
+						return Err(SecondingError::ParentHeadDataMismatch)
+					}
+					pvd
+				},
+				(Some(pvd), Some(parent_head)) => {
+					if parent_head.hash() != parent_head_data_hash {
+						return Err(SecondingError::ParentHeadDataMismatch)
+					}
+					pvd
+				},
+				(None, _) => return Err(SecondingError::PersistedValidationDataNotFound),
+			};
```

Comment:
Arent `Some` cases identical in validation logic ?

```suggestion
let pvd = match maybe_pvd {
    None if maybe_parent_head_data.is_none() => 
        return Err(SecondingError::BlockedOnParent(parent_head_data_hash)),
    None => return Err(SecondingError::PersistedValidationDataNotFound),
    Some(pvd) => {
        let expected_hash = maybe_parent_head_data
            .as_ref()
            .map(|h| h.hash())
            .unwrap_or_else(|| pvd.parent_head.hash());
        if expected_hash != parent_head_data_hash {
            return Err(SecondingError::ParentHeadDataMismatch)
        }
        pvd
    },
};
```